src/replayq_mem_ets_shared.erl

```diff
@@ -0,0 +1,168 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc Shared ETS table for in-memory queue.
+%% Each item in the queue is a tuple of the form `{Id, Item}`.
+%% where `Id` is a tuple of `{self(), NextId}` and `NextId` is an increasing integer.
+%% Each owner process can only create one queue.
+-module(replayq_mem_ets_shared).
+
+-behaviour(replayq_mem).
+
+-export([
+    boot_init/0,
+    purge_by_owner/1
+]).
+
+-export([
+    new/1,
+    peek_all/1,
+    peek/1,
+    is_empty/1,
+    in/2,
+    in_r/2,
+    in_batch/2,
+    out/1,
+    destroy/1
+]).
+
+-export_type([queue/0]).
+
+%% parse-transform for ets:fun2ms/1
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-define(ETS_TAB, ?MODULE).
+-define(KEY(PID, ID), {PID, ID}).
+
+-type key() :: {pid(), non_neg_integer()}.
```

Comment:
Nit: could be negative, if `in_r` is called enough times.

Reply:
right.  with this proposed change, then I think the behavior would be correct even with generic usage of this module: https://github.com/emqx/replayq/pull/25#discussion_r1964128259

Reply:
the 3 flavors are only for internal use.
`replayq` cannot provide `in_r` API because of the files on disk nature (unless we write some inverse-order files).

Reply:
still, I think that with a simple change we can avoid the need to raise errors and keep correct behavior

but since it's not exposed, should be fine as well...

Reply:
changed to use minus one billion as the practical impossible ID

do not want to use a tuple (RAM overhead) just for the sake of something impossible to happen.